MSB1025: An internal failure occurred while running MSBuild on Nano Server

Copied from https://github.com/dotnet/cli/issues/4469
## Steps to reproduce

Build the following Windows Container Dockerfile via `docker build -t preview3 .`

```
FROM microsoft/nanoserver:10.0.14393.321

# Install .NET Core SDK
ENV DOTNET_SDK_VERSION 1.0.0-preview3-003892
ENV DOTNET_SDK_DOWNLOAD_URL https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$DOTNET_SDK_VERSION/dotnet-dev-win-x64.$DOTNET_SDK_VERSION.zip

RUN powershell -NoProfile -Command \
        $ErrorActionPreference = 'Stop'; \
        Invoke-WebRequest %DOTNET_SDK_DOWNLOAD_URL% -OutFile dotnet.zip; \
        Expand-Archive dotnet.zip -DestinationPath '%ProgramFiles%\dotnet'; \
        Remove-Item -Force dotnet.zip

RUN setx /M PATH "%PATH%;%ProgramFiles%\dotnet"

# Trigger the population of the local package cache
ENV NUGET_XMLDOC_MODE skip
RUN mkdir warmup \
    && cd warmup \
    && dotnet new \
    && cd .. \
    && rmdir /q/s warmup
```
## Expected  behavior

Dockerfile builds without issues.
## Actual behavior

Dockfile fails to build

```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
System.EntryPointNotFoundException: Unable to find an entry point named 'RtlMoveMemory' in DLL 'kernel32.dll'.
   at Microsoft.Build.Internal.CommunicationsUtilities.CopyMemory(Char* destination, Char* source, UInt32 length)
   at Microsoft.Build.Internal.CommunicationsUtilities.GetEnvironmentCharArray()
   at Microsoft.Build.Internal.CommunicationsUtilities.GetEnvironmentVariables()
   at Microsoft.Build.Internal.Utilities.GetEnvironmentProperties()
   at Microsoft.Build.Evaluation.ProjectCollection.get_EnvironmentProperties()
   at Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection()
   at Microsoft.Build.Evaluation.ProjectCollection..ctor(IDictionary`2 globalProperties, IEnumerable`1 loggers, IEnumerable`1 remoteLoggers, ToolsetDefinitionLocations toolsetDefinitionLocations, Int32 maxNodeCount, Boolean onlyLogCriticalEvents)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, Boolean debugger, Boolean detailedSummary)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)

Unhandled Exception: System.EntryPointNotFoundException: Unable to find an entry point named 'RtlMoveMemory' in DLL 'kernel32.dll'.
   at Microsoft.Build.Internal.CommunicationsUtilities.CopyMemory(Char* destination, Char* source, UInt32 length)
   at Microsoft.Build.Internal.CommunicationsUtilities.GetEnvironmentCharArray()
   at Microsoft.Build.Internal.CommunicationsUtilities.GetEnvironmentVariables()
   at Microsoft.Build.Internal.Utilities.GetEnvironmentProperties()
   at Microsoft.Build.Evaluation.ProjectCollection.get_EnvironmentProperties()
   at Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection()
   at Microsoft.Build.Evaluation.ProjectCollection..ctor(IDictionary`2 globalProperties, IEnumerable`1 loggers, IEnumerable`1 remoteLoggers, ToolsetDefinitionLocations toolsetDefinitionLocations, Int32 maxNodeCount, Boolean onlyLogCriticalEvents)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, Boolean debugger, Boolean detailedSummary)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)
Failed to create prime the NuGet cache. restore3 failed with: 255
```
## Environment data

I am using a WS16 environment with the latest updates.

This issue was caught by the dotnet-docker-nightly repo's CI which automatically picks up the latest CLI builds
[PR to uptake 003892](https://github.com/dotnet/dotnet-docker-nightly/pull/78)
[CI failure](https://ci.dot.net/job/dotnet_dotnet-docker-nightly/job/master/job/NanoServer_prtest/10/console)

Well, one layer of the onion is removed--it fails differently now:

```
C:\testsapp>dotnet build
Microsoft (R) Build Engine version 15.1.451.4340
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018: The "ResolveAssemblyReference" task failed unexpectedly.\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018: System.DllNotFoundException: Unable to load DLL 'mscoree.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.NativeMethods.GetFileVersion(String szFullPath, StringBuilder szBuffer, Int32 cchBuffer, UInt32& dwLength)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.AssemblyInformation.GetRuntimeVersion(String path)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.AssemblyInformation.IsWinMDFile(String fullPath, GetAssemblyRuntimeVersion getAssemblyRuntimeVersion, FileExists fileExists, String& imageRuntimeVersion, Boolean& isManagedWinmd)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.Reference.set_FullPath(String value)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ReferenceTable.ResolveReference(AssemblyNameExtension assemblyName, String rawFileNameCandidate, Reference reference)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ReferenceTable.SetPrimaryAssemblyReferenceItem(ITaskItem referenceAssemblyName)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ReferenceTable.SetPrimaryItems(ITaskItem[] referenceAssemblyFiles, ITaskItem[] referenceAssemblyNames, ArrayList exceptions)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ReferenceTable.ComputeClosure(DependentAssembly[] remappedAssembliesValue, ITaskItem[] referenceAssemblyFiles, ITaskItem[] referenceAssemblyNames, ArrayList exceptions)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ResolveAssemblyReference.Execute(FileExists fileExists, DirectoryExists directoryExists, GetDirectories getDirectories, GetAssemblyName getAssemblyName, GetAssemblyMetadata getAssemblyMetadata, GetLastWriteTime getLastWriteTime, GetAssemblyRuntimeVersion getRuntimeVersion, GetAssemblyPathInGac getAssemblyPathInGac, IsWinMDFile isWinMDFile, ReadMachineTypeFromPEHeader readMachineTypeFromPEHeader)\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.Tasks.ResolveAssemblyReference.Execute()\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()\r [C:\testsapp\testsapp.csproj]
C:\Program Files\dotnet\sdk\1.0.0-preview4-004176\Microsoft.Common.CurrentVersion.targets(1909,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__25.MoveNext() [C:\testsapp\testsapp.csproj]
```

--- a/dir.props
+++ b/dir.props
@@ -346,6 +346,7 @@
     <DefineConstants>$(DefineConstants);FEATURE_WIN32_REGISTRY</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_WORKINGSET</DefineConstants>
     <DefineConstants Condition="'$(MonoBuild)' != 'true'">$(DefineConstants);FEATURE_VISUALSTUDIOSETUP</DefineConstants>
+    <DefineConstants>$(DefineConstants);FEATURE_MSCOREE</DefineConstants>
   </PropertyGroup>
 
   <PropertyGroup>

--- a/src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs
+++ b/src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs
@@ -457,6 +457,7 @@ namespace Microsoft.Build.Tasks
         /// <returns>The CLR runtime version or empty if the path does not exist.</returns>
         internal static string GetRuntimeVersion(string path)
         {
+#if FEATURE_MSCOREE
             if (NativeMethodsShared.IsWindows)
             {
                 StringBuilder runtimeVersion = null;
@@ -489,6 +490,9 @@ namespace Microsoft.Build.Tasks
             {
                 return ManagedRuntimeVersionReader.GetRuntimeVersion(path);
             }
+#else
+                return ManagedRuntimeVersionReader.GetRuntimeVersion(path);
+#endif
         }
 
 

--- a/src/XMakeTasks/NativeMethods.cs
+++ b/src/XMakeTasks/NativeMethods.cs
@@ -1304,6 +1304,7 @@ typedef enum _tagAssemblyComparisonResult
         [return: MarshalAs(UnmanagedType.Bool)]
         internal static extern bool CertFreeCertificateContext(IntPtr CertContext);
 
+#if FEATURE_MSCOREE
         /// <summary>
         /// Get the runtime version for a given file
         /// </summary>
@@ -1314,6 +1315,7 @@ typedef enum _tagAssemblyComparisonResult
         /// <returns>HResult</returns>
         [DllImport(MscoreeDLL, SetLastError = true, CharSet = CharSet.Unicode)]
         internal static extern uint GetFileVersion(String szFullPath, StringBuilder szBuffer, int cchBuffer, out uint dwLength);
+#endif
         #endregion
 
         #region Methods

